AppKit/CPStringDrawing.j

```diff
@@ -53,20 +57,50 @@ CPStringSizeCachingEnabled = YES;
     return [self sizeWithFont:aFont inWidth:NULL];
 }
 
+
 - (CGSize)sizeWithFont:(CPFont)aFont inWidth:(float)aWidth
 {
     if (!CPStringSizeCachingEnabled)
         return [CPPlatformString sizeOfString:self withFont:aFont forWidth:aWidth];
 
-    var cacheKey = self + [aFont cssString] + aWidth,
+    var cssString = [aFont cssString],
+        cacheKey = self + cssString + aWidth,
         size = CPStringSizeWithFontInWidthCache[cacheKey];
 
-    if (size === undefined)
+    if (size !== undefined)
+        return CGSizeMakeCopy(size);
+
+#if PLATFORM(DOM)
+    if (!CPStringSizeMeasuringContext)
+        CPStringSizeMeasuringContext = CGBitmapGraphicsContextCreate();
+
+    if (!CPStringSizeDidTestCanvasSizingValid && CPFeatureIsCompatible(CPHTMLCanvasFeature))
     {
+        var teststring = "0123456879abcdefghiklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ,.-()";
+        CPStringSizeDidTestCanvasSizingValid = YES;
+        CPStringSizeMeasuringContext.font = cssString;
+        CPStringSizeIsCanvasSizingInvalid = ABS([teststring sizeWithFont:aFont].width - CPStringSizeMeasuringContext.measureText(teststring).width) > 2;
+    }
+
```

Comment:
im not sure whether this is a good idea, since we are in a category to CPString. If we implement +initialize here, we might introduce a conflict to a +initialize in the CPString main file (even if this initializer it does not exist yet, it may be introduced in the future and then break the string sizing stuff).